Core/clim-basic/extended-streams/stream-input.lisp

```diff
@@ -43,6 +44,104 @@
               (eql modifiers +shift-key+))
       (keyboard-event-character event))))
 
+
+(deftype input-gesture-object ()
+  '(or character event))
+
+;;; CLIM's input kernel is only hinted in the specification. -- jd 2019-06-24
+(defclass input-stream-kernel (standard-sheet-input-mixin)
+  ((input-buffer :initarg :input-buffer :accessor stream-input-buffer))
+  (:default-initargs
+   :input-buffer (or *input-buffer* (make-instance 'concurrent-event-queue))))
+
+(defgeneric stream-append-gesture (stream gesture)
+  (:method ((stream input-stream-kernel) gesture)
+    (event-queue-append (stream-input-buffer stream) gesture)))
+
+;;; This function is like "peek-no-hang" but only locally in the buffer, i.e it
+;;; does not trigger wait on event-queue.
+(defgeneric stream-gesture-available-p (stream)
+  (:method ((stream input-stream-kernel))
+    (event-queue-peek (stream-input-buffer stream))))
+
+;;; Default input-stream-kernel methods.
+
+;;; The input buffer is expected to be filled from HANDLE-EVENT methods. All
+;;; tests operate on the stream's input buffer. -- jd 2020-07-28
+(defmethod stream-input-wait ((stream input-stream-kernel) &key timeout input-wait-test)
+  (loop
+    with wait-fun = (and input-wait-test (curry input-wait-test stream))
+    with timeout-time = (and timeout (+ timeout (now)))
+    when (stream-gesture-available-p stream)
+      do (return-from stream-input-wait t)
+    do (multiple-value-bind (available reason)
+           (event-listen-or-wait stream :timeout timeout
+                                        :wait-function wait-fun)
+         (when (and (null available) (eq reason :timeout))
+           (return-from stream-input-wait (values nil :timeout)))
+         (when-let ((event (event-read-no-hang stream)))
+           (handle-event (event-sheet event) event))
+         (when timeout
+           (setf timeout (compute-decay timeout-time nil)))
+         (when (funcall input-wait-test stream)
+           (return-from stream-input-wait
+             (values nil :input-wait-test))))))
```

Comment:
I think that this would be quite intrusive change, so I'd rather not rename it -- see above comment. In this function everything is correct: no input buffer is accessed, stream-input-wait accesses *the* event-queue to advance processing, input-buffer is filled from handle-event methods (via stream-append-gesture).